lib/gelf/ruby_sender.rb

```diff
@@ -22,30 +28,81 @@ def close
     end
   end
 
+  # TCP/TLS socket management
   class RubyTcpSocket
     attr_accessor :socket
 
+    require "socket"
+    require "openssl"
+    require "timeout"
+
     def initialize(host, port)
       @host = host
       @port = port
+      @options = Notifier.options
       connect
     end
 
     def connected?
       if not @connected
-        begin
-          if @socket.nil?
-            @socket = Socket.new(Socket::AF_INET, Socket::SOCK_STREAM, 0)
+        if defined? @options['tls'] and @options['tls'] != true
+          begin
+            if @socket.nil?
+              @socket = Socket.new(Socket::AF_INET, Socket::SOCK_STREAM, 0)
+            end
+            sockaddr = Socket.sockaddr_in(@port, @host)
+            @socket.connect_nonblock(sockaddr)
+          rescue Errno::EISCONN
+            @connected = true
+          rescue Errno::EINPROGRESS, Errno::EALREADY
+            @connected = false
+          rescue SystemCallError
+            @socket = nil
+            @connected = false
+          end
+        else
+          begin
+            if @tcp.nil?
+              @tcp = Socket.new(
+                Socket::Constants::AF_INET,
+                Socket::Constants::SOCK_STREAM,
+                Socket::Constants::IPPROTO_IP
+              )
+              @tcp = TCPSocket.new @host, @port
+            end
+            if @socket.nil?
+              tls_context = OpenSSL::SSL::SSLContext.new
+              unless @options['check_ssl'].nil?
+                if @options['check_ssl'] == true
+                  tls_context.set_params({ :verify_mode=>OpenSSL::SSL::VERIFY_PEER})
+                else
+                  tls_context.set_params({ :verify_mode=>OpenSSL::SSL::VERIFY_NONE})
+                end
+              end
+              unless @options['tls_version'].nil?
+                if OpenSSL::SSL::SSLContext::METHODS.any? { |v| v.to_s.include?(@options['tls_version']) }
+                  tls_context.set_params({ :ssl_version => @options['tls_version']})
+                end
+              end
+              @socket = OpenSSL::SSL::SSLSocket.new(@tcp,tls_context)
+              @socket.sync_close = true
+              @socket.connect
+            end
+            @connected = true
+          rescue Errno::EISCONN
+            @connected = true
+          rescue Errno::EINPROGRESS, Errno::EALREADY
+            @connected = false
+          rescue Errno::EHOSTUNREACH, Errno::ENETUNREACH, Errno::ECONNABORTED, Errno::ECONNREFUSED, Errno::ETIMEDOUT
+            @socket = nil
+            @connected = false
+          rescue OpenSSL::SSL::SSLError
+            @socket = nil
+            @connected = false
+          rescue SystemCallError
```

Comment:
I came across the following error while testing:

```
/usr/lib/ruby/2.1.0/openssl/ssl.rb:130:in `fcntl': closed stream (IOError)
	from /usr/lib/ruby/2.1.0/openssl/ssl.rb:130:in `initialize'
	from /home/bernd/projects/gelf-rb/lib/gelf/ruby_sender.rb:87:in `initialize'
	from /home/bernd/projects/gelf-rb/lib/gelf/ruby_sender.rb:87:in `new'
	from /home/bernd/projects/gelf-rb/lib/gelf/ruby_sender.rb:87:in `connected?'
	from /home/bernd/projects/gelf-rb/lib/gelf/ruby_sender.rb:225:in `block in send'
	from /home/bernd/projects/gelf-rb/lib/gelf/ruby_sender.rb:224:in `map'
	from /home/bernd/projects/gelf-rb/lib/gelf/ruby_sender.rb:224:in `send'
	from /home/bernd/projects/gelf-rb/lib/gelf/notifier.rb:169:in `notify_with_level!'
	from /home/bernd/projects/gelf-rb/lib/gelf/notifier.rb:142:in `notify!'
```

Can you please add another rescue claus to handle this:

```diff
--- lib/gelf/ruby_sender.rb
+++ lib/gelf/ruby_sender.rb
@@ -102,6 +102,9 @@ module GELF
           rescue SystemCallError
             @socket = nil
             @connected = false
+          rescue IOError
+            @socket = nil
+            @connected = false
           end
         end
       end
```